File: tests/weather/normalize.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import {
  normalizeAirQuality,
  normalizeCurrent,
  normalizeForecastPoints,
  normalizeLocation,
} from "../../src/weather/openweather/normalize";

const dir = join(import.meta.dirname, "../fixtures/openweather");

function load(name: string): unknown {
  return JSON.parse(readFileSync(join(dir, name), "utf8"));
}

describe("normalizeCurrent", () => {
  it("maps a full current-weather payload into application models", () => {
    const current = normalizeCurrent(load("current.json"));
    const location = normalizeLocation(load("current.json"), "Fallback");
    expect(location.displayName).toBe("Stockholm");
    expect(location.country).toBe("SE");
    expect(location.coordinates).toEqual({ latitude: 59.3293, longitude: 18.0686 });
    expect(location.timezoneOffsetSeconds).toBe(7200);
    expect(current.temperatureC).toBe(14.2);
    expect(current.feelsLikeC).toBe(13.1);
    expect(current.highC).toBe(15.4);
    expect(current.lowC).toBe(12.8);
    expect(current.condition).toBe("cloudy");
    expect(current.conditionText).toBe("broken clouds");
    expect(current.humidityPercent).toBe(72);
    expect(current.pressureHpa).toBe(1014);
    expect(current.visibilityM).toBe(10000);
    expect(current.cloudPercent).toBe(75);
    expect(current.wind).toEqual({ speedMps: 4.1, directionDeg: 240, gustMps: 7.2 });
    expect(current.precipitation.rainMm).toBe(0.2);
    expect(current.precipitation.probabilityPercent).toBeUndefined();
    expect(current.observedAtMs).toBe(1724260800 * 1000);
  });

  it("keeps omitted precipitation distinct from zero", () => {
    const current = normalizeCurrent(load("current-sparse.json"));
    expect(current.precipitation.rainMm).toBeUndefined();
    expect(current.precipitation.snowMm).toBeUndefined();
    expect(current.highC).toBeUndefined();
    expect(current.lowC).toBeUndefined();
    expect(current.visibilityM).toBeUndefined();
    expect(current.wind.gustMps).toBeUndefined();
  });

  it("rejects structurally invalid current weather", () => {
    expect(() => normalizeCurrent({})).toThrowError(/missing/i);
  });
});

describe("normalizeForecastPoints", () => {
  it("skips malformed points without dropping the rest", () => {
    const points = normalizeForecastPoints(load("forecast.json"));
    expect(points).toHaveLength(3);
    expect(points[0]?.condition).toBe("light-rain");
    expect(points[0]?.precipitation.probabilityPercent).toBe(40);
    expect(points[0]?.precipitation.rainMm).toBe(0.31);
    expect(points[1]?.precipitation.rainMm).toBeUndefined();
    expect(points[2]?.precipitation.probabilityPercent).toBe(0);
  });

  it("fails when no usable points remain", () => {
    expect(() =>
      normalizeForecastPoints({ list: [{ dt: "nope" }] }),
    ).toThrowError(/no usable/i);
  });
});

describe("normalizeAirQuality", () => {
  it("preserves OpenWeather AQI rather than renaming it EPA", () => {
    const air = normalizeAirQuality(load("air.json"));
    expect(air.openWeatherAqi).toBe(2);
    expect(air.category).toBe("fair");
    expect(air.components.pm2_5).toBe(4.12);
  });
});

File: app/src/test/java/org/radilabs/weather/weather/openweather/NormalizeTest.kt
package org.radilabs.weather.weather.openweather

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.radilabs.weather.weather.ConditionCategory
import org.radilabs.weather.weather.WeatherError

class NormalizeTest {
    @Test
    fun currentMapsMetricUnitsAndCondition() {
        val current = normalizeCurrent(parseObject(readFixture("current.json"), "current"))
        assertEquals(14.2, current.temperatureC, 0.0)
        assertEquals(13.1, current.feelsLikeC, 0.0)
        assertEquals(4.1, current.wind.speedMps, 0.0)
        assertEquals(0.2, current.precipitation.rainMm)
        assertEquals(ConditionCategory.Cloudy, current.condition)
        assertEquals("broken clouds", current.conditionText)
        assertEquals(1_724_260_800_000L, current.observedAtMs)
        assertEquals(1014.0, current.pressureHpa)
        assertEquals(10000.0, current.visibilityM)
    }

    @Test
    fun forecastSkipsMalformedPoints() {
        val points = normalizeForecastPoints(parseObject(readFixture("forecast.json"), "forecast"))
        assertEquals(3, points.size)
        assertEquals(ConditionCategory.LightRain, points[0].condition)
        assertEquals(40.0, points[0].precipitation.probabilityPercent)
        assertEquals(0.31, points[0].precipitation.rainMm)
        assertEquals(0.0, points[2].precipitation.probabilityPercent)
    }

    @Test
    fun airQualityIsOpenWeatherIndex() {
        val air = normalizeAirQuality(parseObject(readFixture("air.json"), "air"))
        assertEquals(2, air.openWeatherAqi)
        assertEquals(org.radilabs.weather.weather.AirQualityCategory.Fair, air.category)
        assertEquals(4.12, air.components.pm25)
    }

    @Test
    fun malformedCurrentThrows() {
        try {
            normalizeCurrent(parseObject("{\"main\":{}}", "current"))
            throw AssertionError("expected malformed")
        } catch (error: WeatherError) {
            assertEquals(WeatherError.Code.Malformed, error.code)
        }
    }

    private fun readFixture(name: String): String {
        val stream = requireNotNull(javaClass.classLoader?.getResourceAsStream("openweather/$name")) {
            "missing fixture $name"
        }
        return stream.bufferedReader().readText()
    }
}
